### PR TITLE
Output Guardrails

### PR DESCRIPTION
**Pull request by chaecramb**

### CODE REVIEW

File: tests/output_guardrails/test_generate.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+def test_generate_models_to_evaluation_models_returns_evaluation_results(
+    run_rake_task_mock,
+):
+    def result_per_question(_, env):
+        if env["INPUT"] == "Question 1":
+            return {"triggered": True, "llm_guardrail_result": 'True | "1, 4, 7"'}
+        else:
+            return {"triggered": False, "llm_guardrail_result": "False | None"}
+
+    run_rake_task_mock.side_effect = result_per_question
```
Comment (kevindew) [inline tests/output_guardrails/test_generate.py:39]: I don't think you need this because a side effect is already configured in the mock?

File: tests/output_guardrails/conftest.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    ]
+
+    path = mock_project_root / "input_data.jsonl"
+
+    with open(path, "w") as file:
+        for item in data:
+            json.dump(item, file)
+            file.write("\n")
+
+    return str(path)
```
Comment (kevindew) [inline tests/output_guardrails/conftest.py:32]: I might be wrong, but I don't think we need to convert this to a string from a path (I imagine it might be something to fix in existing code)

File: tests/output_guardrails/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        "what": "Testing Ooutput Guardrail evaluations",
+        "guardrail_type": "answer_guardrails",
+        "generate": False,
+        "input_path": str(mock_input_data),
+    }
+    file_path = tmp_path / "config.yaml"
+    with open(file_path, "w") as file:
+        yaml.dump(data, file)
+
+    yield str(file_path)
```
Comment (kevindew) [inline tests/output_guardrails/test_cli.py:50]: As per previous comment, I think (but could be wrong) that we don't need to convert this path to a string (and the one a few lines up)

File: tests/output_guardrails/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        )
+
+
+@pytest.fixture(autouse=True)
+def mock_config_file(tmp_path, mock_input_data):
+    """Write a config file as an input for testing"""
+    data = {
+        "what": "Testing Ooutput Guardrail evaluations",
+        "guardrail_type": "answer_guardrails",
+        "generate": False,
```
Comment (kevindew) [inline tests/output_guardrails/test_cli.py:43]:
```suggestion
        "generate": True,
```

I think we should probably set this to true for the tests, it seems to create a more honest scenario for no-generate (another thing to fix in jailbreak guardrails)

File: tests/output_guardrails/test_cli.py (hunk 3 above)
Comment (chaecramb) [reply tests/output_guardrails/test_cli.py:50]: The previous one is fine to remove, but these two blows up the tests, I think because Click's CLI runner requires a string path

Comment (kevindew) [reply tests/output_guardrails/test_cli.py:51]: Great - makes sense

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def _calculate_metric(self, numerator: int, denominator: int) -> float:
+        if denominator == 0:
+            return float("nan")
+        return float(numerator) / denominator
+
+    def precision_exact(self) -> float:
+        return self._calculate_metric(
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:101]: It is a bit surprising that we have two different techniques in this class to calculate precision and recall. I'm expecting we went with this different technique because it wasn't as easy to create the lists.

Assuming it is a bit of a pain to create the lists and thus use sklearn maybe we want to just use the helper function. 

In doing so I think we should rename from the helper of "_calculate_metric" as that is very non specific (and really is just a division function) to a `_calculate_precision(true_positives, false_positives)` and `_calculate_recall(true_positives, false_negatives)`

**Review by kevindew — COMMENTED** [review]
Great job here, just a few more comments now I've had a chance to review more.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                return "true_positive"
+            case (False, False):
+                return "true_negative"
+            case (False, True):
+                return "false_positive"
+            case (True, False):
+                return "false_negative"
+
+    @property
+    def classification_exact(self) -> str:
```
[the remark this answers is not in the document]
Comment (exfalsoquodlibet) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:34]: Hi @chaecramb @kevindew, I gave this some thinking after our chat, Kev. 

This is a case of a multi-label classification problem, in that a case (a test answer) can be assigned to more than one non-exclusive labels (more than one guardrail triggers). In the output guardrails, there are 7 such guardrails (or criteria in a certain sense) for both the RAG output and the question-router output:
- https://github.com/alphagov/govuk_chat_private/blob/56e7cb4408e137af64ce7c0d36628512c6209eb6/config/llm_prompts/openai/guardrails.yml#L139-L146
- https://github.com/alphagov/govuk_chat_private/blob/56e7cb4408e137af64ce7c0d36628512c6209eb6/config/llm_prompts/openai/guardrails.yml#L167-L174

I was not too close to the implementation of the output guardrails or its original evaluation, but this is, I think, what we need.

For the "exact evaluation", we want to be able to **understand how each individual criterion/guardrail is performing**, so that we can identify the ones that are problematic and iterate on them, while also recording how these changes may affect the classification of teh other criteria.

So, assuming with we have three test cases as strings of indexes, where each index is one of 1:7. one for each criterion.

```
ground_truth_exacts = ["1,3", "2", "4,5"]
predicted_exacts = ["1,3", "2,4", "5"]
labels=7
```

So we can see:
- first prediction: perfect match
- second prediction: got one right "2", but incorrectly triggered also "4"
- third prediction: got one right "5" but missed one "4"

When looking at the individual critera:
- guardrail criteria "1": 1 TP
- guardrail criteria "2": 1 TP
- guardrail criteria "3": 1 TP
- guardrail criteria "4": 1 TN and 1 FP
- guardrail criteria "5": 1 TP
- guardrail criteria "6": -
- guardrail criteria "7": - 

Let's convert these cases into binary vectors so we can do some maths on them. For each case, we are going to have a vector with 7 indexes, where each index corresponds to one of the 7 guardrail criteria. An index is assigned 0, unless that guardrail criteria was triggered (or should be triggered, in case of the ground-truth), in which case it gets a 1. So we have:

```
ground_truth_vectors = [
  [1, 0, 1, 0, 0, 0, 0],
  [0, 1, 0, 0, 0, 0, 0],
  [0, 0, 0, 1, 1, 0, 0],
]

predicted_vectors = [
  [1, 0, 1, 0, 0, 0, 0],
  [0, 1, 0, 1, 0, 0, 0],
  [0, 0, 0, 0, 1, 0, 0],
]
```

A quick python function to do this conversaion:
```
def str_to_vector(label_str: str, n_labels: int):
     vec = [0] * n_labels
     if label_str.strip == "":
           return vec
     for idx in label_str.split(','):
           vec[int(idx)-1] = 1
     return vec  

ground_truth_vectors = [str_to_vector(case, labels) for case in ground_truth_exacts]
# etc..

# we may want to handle cases in case the LLM hallucinate numbers outside the n_labels range, not sure whether you have seen such cases
```


Now each predicted case and its ground-truth reference is a vector of integers, and we can compare them. 

There several ways to do this comparison. The simplest one, but too harsh, would be to see whether the ground-truth and the predicted vector are a perfect match. 

My suggestion is that:
- we calculate per-label (so per-criterion) metrics of recall and precision

To calculate these metrics, we can `skit-learn`:
```
from sklearn.metrics import precision_score, recall_score, f1_score

precision_per_guardrail = precision_score(ground_truth_vectors, predicted_vectors, average=None, zero_division=0)

recall_per_guardrail = recall_score(ground_truth_vectors, predicted_vectors, average=None, zero_division=0)

f1_per_guardrail = f1_score(ground_truth_vectors, predicted_vectors, average=None, zero_division=0)

# can collect and output as a pandas dataframe

pd.DataFrame({
   "precision" : precision_per_guardrail,
    "recall": recall_per_guardrail,
    "f1 score": f1_per_guardrail
```
So we will have an array of precision, an array recall and an array of f1-scores, one score per guardrail criterion.

I think this approach would give us the information we need to run diagnostic on the separate guardrail criteria, so which guardrail criterion performs well, which one struggles, help us priortise improvements, etc.

We could also report an average (across all criteria), in which case I suggest we go for "macro" averages. But i believe the metric breakdown by guardrail criterion is what we need.

I hope it makes sense, happy to chat!

[the remark this answers is not in the document]
Comment (kevindew) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:34]: Thanks Alessia, I think that makes a lot of sense. I imagine we want to see if there’s a relatively clear route to achieving that in this PR or whether to break it out into another PR, though given the exact stuff is a bit contrived it seems good to do as part of this story.

So I imagine what we could do is have two methods on EvaluationResult that return vectors: expected_guardrails_triggered -> list[int] and actual_guardrails_triggered -> list[int], both parse the expected_exact and actual_exact strings to build those lists and raise errors if they don’t match formatting or have a guardrail number greater than 7.

Aside: I do feel a bit uneasy about the hardcoding of 7, it feels rather fragile - I imagine it could be more robust of the response from the rake task had the actual max number in it - but that could be a future todo. Better still could be that the rake task returns a mapping of number to guardrail name so then these results could include both the number of the guardrail and a more human recognisable name.

So all that part seems quite achievable

— 

I guess from there we’d just need to work out a relatively quick way to decide what to dump as a csv file or include in the aggregate values. 

@chaecramb what's your thoughts? I imagine we might be best to cut back the exact stuff in this and then open a new PR to investigate this bit and how it impacts the aggregate report.

[the remark this answers is not in the document]
Comment (kevindew) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:34]: Sure, both suggestions would involve returning the configuration data of the guardrails and not be relying on the LLM

[the remark this answers is not in the document]
Comment (chaecramb) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:34]: @kevindew Yeah, given this PR was just to get the overall structure in, I think I could just strip the 'exact' part out of it. I'm not sure there's a benefit to including the current contrived version only to immediately change it. 

If I rebase this PR to just include the 'triggered' version, so that it matches the jailbreak guardrails approach, then I think that provides a basic structure for a follow up PR that implements an 'exact' evaluation in a way that's actually useful, following @exfalsoquodlibet's explanation above.

**Comment by chaecramb** [comment]
@kevindew I've stripped this back to just do the triggered classification for now. I'll start a follow up PR for implementing Alessia's explanation of multiple-label classification from above for the exact approach.

**Review by kevindew — APPROVED** [review]
Nice one
